File: No-lineal structure/Structure in js/Arboles/Merkle/Block.js
import { MerkleTree } from "./merkle.js";

export class Data {
    constructor(_from = null, skins = null) {
        this._from = _from;
        this.skins = skins;
    }
}

export class Block{
    difficulty_target = "0000";
    index;
    timestamp;
    data;
    nonce;
    previoushash;
    rootmerkle;
    hash;

    constructor(index, data, difficulty_target, previoushash=''){
        this.index = index;
        this.difficulty_target = difficulty_target;
        this.timestamp = datetime.now().strftime("%d-%m-%Y::%H:%M:%S");
        this.data = data;
        this.nonce = 0;
        this.previoushash = previoushash;
        this.mtEDD = null; //Se toma la referencia en memoria del arbol merkle
        this.merkleroot = this.calcMerkleRoot();
        this.hash = this.calcBlockhash();

        this.next = null;
        this.back = null;
    }

    calcMerkleRoot() {
        let elems = [];
        for (x in this.data.skins['data']) {
            //print(x)
            elems.append(x["SKIN"]+str(x["VALUE"]));
        }
        this.mtEDD = MerkleTree(elems);
        //mt.graficar(mt.getRootHash())
        this.mtEDD.graficar(str(this.index));
        return this.mtEDD.getRootHash();
    }
    
    calcBlockhash() {
        // Verificar el orden de concatenacion
        let nonce = 0;
        while (true) {
            let h = str(this.index) + str(this.timestamp) + this.previoushash + this.merkleroot + str(nonce);
            nonce += 1;
            if (str(hashlib.sha256(h.encode('utf-8')).hexdigest()).find(this.difficulty_target) == 0) {
                this.nonce = nonce;
                return hashlib.sha256(h.encode('utf-8')).hexdigest();
            }
        }
        //return hashlib.sha256(h.encode('utf-8')).hexdigest()
    }

    rawblockInfo() {
        return {
            index: str(self.index),
            timestamp: self.timestamp,
            nonce: str(self.nonce),
            data: {
                FROM: self.data._from,
                SKINS: self.data.skins
            },
            hash: self.hash,
            previoushash: self.previoushash,
            merkle_root: self.merkleroot
        };
    }
    
    blockInfo() {
        return JSON.stringify(rawblockInfo(), null, 4);
    }
}

File: No-lineal structure/Structure in js/MerkleBlocchain/Block.py
from datetime import datetime
import hashlib
from Merkle import MerkleTree
import json

class Data:
    _from: str
    skins: list

    def __init__(self, _from=None, skins=None) -> None:
        self._from = _from
        self.skins = skins

class Block:
    difficulty_target = "0000"
    index: int
    timestamp: str
    data: Data
    nonce: int
    previoushash: str
    rootmerkle: str
    hash: str

    def __init__(self, index, data, difficulty_target, previoushash=''):
        self.index = index
        self.difficulty_target = difficulty_target
        self.timestamp = datetime.now().strftime("%d-%m-%Y::%H:%M:%S")
        self.data: Data = data
        self.nonce = 0
        self.previoushash = previoushash
        self.mtEDD = None #Se toma la referencia en memoria del arbol merkle
        self.merkleroot = self.calcMerkleRoot()
        self.hash = self.calcBlockhash()

        self.next = None
        self.back = None

    def calcMerkleRoot(self)-> str:
        elems = []
        for x in self.data.skins['data']:
            #print(x)
            elems.append(x["SKIN"]+str(x["VALUE"]))
        self.mtEDD = MerkleTree(elems)
        #mt.graficar(mt.getRootHash())
        self.mtEDD.graficar(str(self.index))
        return self.mtEDD.getRootHash()

    def calcBlockhash(self):
        # Verificar el orden de concatenacion
        nonce = 0
        while(True):
            h:str = str(self.index) + str(self.timestamp) + self.previoushash + self.merkleroot + str(nonce)
            nonce += 1
            if(str(hashlib.sha256(h.encode('utf-8')).hexdigest()).find(self.difficulty_target) == 0):
                self.nonce = nonce
                return hashlib.sha256(h.encode('utf-8')).hexdigest()
        #return hashlib.sha256(h.encode('utf-8')).hexdigest()

    def rawblockInfo(self):
        return {'index': str(self.index), 'timestamp': self.timestamp, 'nonce': str(self.nonce), "data": {"FROM": self.data._from, "SKINS": self.data.skins} ,'hash': self.hash, 'previoushash': self.previoushash, 'merkle_root': self.merkleroot}

    def blockInfo(self):
        return json.dumps(self.rawblockInfo(), indent=4)